SIL.WritingSystems/Sldr.cs

```diff
@@ -161,21 +161,38 @@ private static void InitializeGetUnicodeCategoryBasedOnIcu()
 
 			// Unfortunately, this version has to be looked up in the docs. There is no way
 			// (e.g. from System.Globalization) to look this up at runtime.
-			const double unicodeVersionOfDotNet4_6_2 = 8.0;
+			const double unicodeVersionOfDotNet462 = 8.0;
 
-			if (double.Parse(Wrapper.UnicodeVersion) > unicodeVersionOfDotNet4_6_2)
-			{
-				StringExtensions.AltImplGetUnicodeCategory = (s, index) =>
-				{
-					if (index < 0 || index >= s.Length)
-						throw new ArgumentOutOfRangeException(nameof(index), index,
-							"Index out of range. Must be a positive number less than the" +
-							$" length ({s.Length}) of the string: {{s}}");
-
-					return Character.GetCharType(char.ConvertToUtf32(s, index))
-						.ToUnicodeCategory();
-				};
-			}
+			if (double.Parse(Wrapper.UnicodeVersion) > unicodeVersionOfDotNet462)
+				StringExtensions.AltImplGetUnicodeCategory = GetUnicodeCategoryBasedOnICU;
+		}
+
+		/// <summary>
+		/// Uses ICU to get the Unicode category of the character at the indicated position in the
+		/// string.
+		/// </summary>
+		/// <exception cref="ArgumentOutOfRangeException">Index out of range. Must be a positive
+		/// number less than the length of the <see cref="s"/></exception>
+		/// <exception cref="ArgumentException">String contains invalid surrogate characters (e.g.,
+		/// a low surrogate that is not preceded by a high surrogate)</exception>
+		/// <remarks>Internal for testing.</remarks>
+		internal static UnicodeCategory GetUnicodeCategoryBasedOnICU(string s, int index)
+		{
+			if (index < 0 || index >= s.Length)
+				throw new ArgumentOutOfRangeException(nameof(index), index,
+					"Index out of range. Must be a positive number less than the" +
+					$" length ({s.Length}) of the string: {{s}}");
+
+			// Just to be a bit more robust (like the default
+			// "UnicodeInfo.GetUnicodeCategory" implementation), if they gave us the index
+			// of the low surrogate and there is a valid high surrogate character preceding
+			// it, we'll fix things up so it doesn't throw an exception.
+			if (char.IsLowSurrogate(s, index) && index > 0 &&
```

Comment:
This won't work on big-endian machines, but I don't think we have to worry about that for libpalaso :smile:

Reply:
Hmmm... Might be worth a comment.

Reply:
Doesn't endian-ness only affect the order of bytes in a character (i.e. the two bytes that make up each character)? I would expect the order of the characters to be unaffected by endian-ness.

Reply:
That's what this is. If the user incorrectly specifies the position between the low and the high surrogate, regardless of endian-ness, ICU considers it an invalid position. On a small-endian implementation, you can detect this (and "fix" the index) by looking to see if the preceding position is the low surrogate and specified one is high. But on a big-endian machine, it will be the opposite.

Reply:
> Doesn't endian-ness only affect the order of bytes in a character (i.e. the two bytes that make up each character)? I would expect the order of the characters to be unaffected by endian-ness.

yes, it affects the order of the bytes in a character, but it also affects the order of the high and low surrogate "characters" - that's why there is a BOM also for UTF-16 and UTF-32 texts.

Reply:
> That's what this is.

No, this method is dealing with characters in a string (each character is two bytes). **index** is the character index, not the byte index.

> but it also affects the order of the high and low surrogate "characters"

Can you point to some documentation about this? I can't find any evidence that endian-ness affects anything but the bytes of each character (UTF-16 and UTF-32 need BOM characters because each character is multi-byte).

Reply:
I think Tim is right: https://unicodebook.readthedocs.io/unicode_encodings.html. Unfortunately, I jumped the gun (twice) and merged this without getting this resolved, and then I pushed a little change to add the comment, which I think now is incorrect. (That one went in without a PR by mistake. I didn't realize that master was not protected in this repo.)

Reply:
@FoolRunning I guess you're right :smile: I got confused by the "high" and "low". The Unicode website also talks about  leading and trailing for high and low surrogates which makes it a bit clearer that the leading/high surrogate always comes first.